anta/tests/avt.py

```diff
@@ -0,0 +1,223 @@
+# Copyright (c) 2023-2024 Arista Networks, Inc.
+# Use of this source code is governed by the Apache License 2.0
+# that can be found in the LICENSE file.
+"""Module related to Adaptive virtual topology tests."""
+
+# Mypy does not understand AntaTest.Input typing
+# mypy: disable-error-code=attr-defined
+from __future__ import annotations
+
+from ipaddress import IPv4Address
+from typing import ClassVar
+
+from pydantic import BaseModel
+
+from anta.models import AntaCommand, AntaTemplate, AntaTest
+from anta.tools import get_value
+
+
+class VerifyAVTPathHealth(AntaTest):
+    """
+    Verifies the status of all Adaptive Virtual Topology (AVT) paths for all VRFs.
+
+    Expected Results
+    ----------------
+    * Success: The test will pass if all AVT paths for all VRFs are active and valid.
+    * Failure: The test will fail if the AVT path is not configured or if any AVT path under any VRF is either inactive or invalid.
+
+    Examples
+    --------
+    ```yaml
+    anta.tests.avt:
+      - VerifyAVTPathHealth:
+    ```
+    """
+
+    name = "VerifyAVTPathHealth"
+    description = "Verifies the status of all AVT paths for all VRFs."
+    categories: ClassVar[list[str]] = ["avt"]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaCommand(command="show adaptive-virtual-topology path")]
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Main test function for VerifyAVTPathHealth."""
+        # Initialize the test result as success
+        self.result.is_success()
+
+        # Get the command output
+        command_output = self.instance_commands[0].json_output.get("vrfs", {})
+
+        # Check if AVT is configured
+        if not command_output:
+            self.result.is_failure("Adaptive virtual topology paths are not configured.")
+            return
+
+        # Iterate over each VRF
+        for vrf, vrf_data in command_output.items():
+            # Iterate over each AVT path
+            for profile, avt_path in vrf_data.get("avts", {}).items():
+                for path, flags in avt_path.get("avtPaths", {}).items():
+                    # Get the status of the AVT path
+                    valid = flags["flags"]["valid"]
+                    active = flags["flags"]["active"]
+
+                    # Check the status of the AVT path
+                    if not valid and not active:
+                        self.result.is_failure(f"AVT path {path} for profile {profile} in VRF {vrf} is invalid and not active.")
+                    elif not valid:
+                        self.result.is_failure(f"AVT path {path} for profile {profile} in VRF {vrf} is invalid.")
+                    elif not active:
+                        self.result.is_failure(f"AVT path {path} for profile {profile} in VRF {vrf} is not active.")
+
+
+class VerifyAVTSpecificPath(AntaTest):
+    """
+    Verifies the status and type of an Adaptive Virtual Topology (AVT) path for a specified VRF.
```

Comment:
So - it is possible to have multiple path with same AVT, VRF, source, next-hop and "type"

```
VRF: default, VNI: 1
AVT: DEFAULT-POLICY-DEFAULT, AVT ID: 1
Application profile(s): default
    Path Type:ID  Destination   Nexthop VTEP   Priority   MTU  Latency   Jitter     Loss Hop List
    ------------- ------------- -------------  -------- ----- -------- -------- -------- ---------------
*>  direct:1      192.168.200.1 192.168.200.1         1  1424 32.946ms 21.226ms  0.0000% path1
*>  direct:2      192.168.200.2 192.168.200.2         1  1424 18.832ms 24.031ms  0.0000% path2
*>  multihop:2564 192.168.200.3 192.168.200.5         1  1424130.783ms 58.995ms  0.0000% path3,8195,3
*>  multihop:2565 192.168.200.3 192.168.200.5         1  1424 87.348ms 30.105ms  0.0000% path3,8198
```

cf the last two from `.3` to `.5` - I guess this test would only check the first one. Let me think a bout this a bit more

Reply:
![Screenshot 2024-04-25 at 10 26 39](https://github.com/arista-netdevops-community/anta/assets/3007422/91ee40b4-77e8-4f95-86a5-1b7a4a53eaca)
ok actually both are picked up so thats fine - we just need to adjust the description saying we would pick up all the paths 

Though if there is a mix of direct and multihop the test will always fail as one path will always be wrong. Can we inside use the type as - this is the type of path we WANT to find- rather than "type is not matching"

so if type is set to `direct` I want to make sure I have a direct path with correct status (and ignore multihop)
if type is `multihop` I want to make sure I have all multihop with correct status (and ignore direct)
if type is not set (so default would be None) - I read all the paths (direct and multihop) and make sure that they are all in correct status

Reply:
Updated
1. If input have path_type as direct we will check only all direct path with mentioned next hop address. if any path is not active of not valid, test will fail. In case no direct path found then also it will fail.
2. If input have path_type as multihop  then will do same as above.
3. If no path_type is defined, we will check all path with mentioned next hop address and will fail if any path is not active or not valid.